packages/cli/src/commands/hydrogen/env/pull.ts

```diff
@@ -0,0 +1,151 @@
+import {Flags} from '@oclif/core';
+import Command from '@shopify/cli-kit/node/base-command';
+import {
+  renderConfirmationPrompt,
+  renderFatalError,
+} from '@shopify/cli-kit/node/ui';
+import {
+  outputContent,
+  outputInfo,
+  outputSuccess,
+  outputToken,
+  outputWarn,
+} from '@shopify/cli-kit/node/output';
+import {fileExists, writeFile} from '@shopify/cli-kit/node/fs';
+import {resolvePath} from '@shopify/cli-kit/node/path';
+
+import {adminRequest, parseGid} from '../../../lib/graphql.js';
+import {commonFlags} from '../../../lib/flags.js';
+import {getHydrogenShop} from '../../../lib/shop.js';
+import {getAdminSession} from '../../../lib/admin-session.js';
+import {
+  PullVariablesQuery,
+  PullVariablesSchema,
+} from '../../../lib/graphql/admin/pull-variables.js';
+import {getConfig} from '../../../lib/shopify-config.js';
+import {hydrogenStorefrontsUrl} from '../../../lib/admin-urls.js';
+
+export default class Pull extends Command {
+  static description =
+    'Populate your .env with variables from your Hydrogen storefront.';
+
+  static hidden = true;
+
+  static flags = {
+    path: commonFlags.path,
+    shop: commonFlags.shop,
+    force: commonFlags.force,
+  };
+
+  async run(): Promise<void> {
+    const {flags} = await this.parse(Pull);
+    await pullVariables(flags);
+  }
+}
+
+interface Flags {
+  force?: boolean;
+  path?: string;
+  shop?: string;
+}
+
+export async function pullVariables({force, path, shop: flagShop}: Flags) {
+  const shop = await getHydrogenShop({path, shop: flagShop});
+  const adminSession = await getAdminSession(shop);
+  const actualPath = path ?? process.cwd();
+  const {storefront: configStorefront} = await getConfig(actualPath);
+
+  if (!configStorefront?.id) {
+    renderFatalError({
+      name: 'NoLinkedStorefrontError',
+      type: 0,
+      message: 'No linked Hydrogen storefront.',
+      tryMessage:
+        outputContent`In order to pull environment variables from a Hydrogen storefront on ${
+          adminSession.storeFqdn
+        } you must first run ${outputToken.genericShellCommand(
+          `npx shopify hydrogen link`,
+        )}.`.value,
+    });
+    return;
+  }
+
+  outputInfo(
+    `Fetching Preview environment variables from ${configStorefront.title}...`,
+  );
+  const result: PullVariablesSchema = await adminRequest(
+    PullVariablesQuery,
+    adminSession,
+    {
+      id: configStorefront.id,
+    },
+  );
```

Comment:
Minor thing: instead of importing the query string and `adminRequest`, perhaps we import a `pullEnvVariables` function that does this internally from `src/lib/...`.

Reply:
But aren't we already doing that? The `PullVariablesQuery` lives in a shared directory and we are importing it here. What I'm saying is using that same file (or around the same place) where PullVariablesQuery is right now, but instead of exporting a raw query, export a function that runs "adminRequest + PullVariablesQuery + adminSession".
But just a minor suggestion, not important!

Reply:
I flipped my opinion on this 😄. As I was working on the `dev` command I saw we'd have to much duplicated logic so I'm working on pulling this out in another branch but we'll leave it as-is in this one for now.